Implement user login

ShouvikGhosh2048 opened the pull request:
Closes #766.

ShouvikGhosh2048 committed:
Basic login

ShouvikGhosh2048 committed:
Add user claims

ShouvikGhosh2048 committed:
Implement verification

ShouvikGhosh2048 committed:
Shift login to top right

ShouvikGhosh2048 commented:
@otobot1 you can review the workflow. I'm left with server logging (and maybe convert the foreign key changes to a single transaction).

ShouvikGhosh2048 commented on prisma/migrations/20240902061032_add_user_claims/migration.sql:
```diff
@@ -0,0 +1,25 @@
+/*
```
I ended up generating a new migration, is this fine or is there a way to merge this with the already present one?

otobot1 replied:
That's intended! There's no way to merge other than deleting existing migrations and restarting our migration history. I might do that so things are cleaner, as the migrations don't really matter at this stage because our production db hasn't had any writes made. Once we start taking new ratings, we'll only ever add migrations.

otobot1 replied:
I generated a new [baseline migration](https://www.prisma.io/docs/orm/prisma-migrate/getting-started#create-a-baseline-migration) to give us a clean migration history before we start accepting user data. For my future reference (and anyone curious :D) here are the steps I used (step 4 differs from the linked guide, as I ran into a bug where Prisma couldn't find my baseline migration at step 5):
1. Delete the **CONTENTS ONLY** of the `_prisma_migrations` table from your local db, if applicable.
2. Create a fresh export of your local db, if applicable.
3. Delete the contents of `prisma/migration`.
4. Run this command: `npx prisma migrate dev --create-only --name 0_init --skip-seed --skip-generate`.
5. Rename the generated migration folder from `YYYYMMDDhhmmss_0_init` to `0_init`.
6. Drop your local `modslist` database, recreate it, and import your fresh export.
7. Run this command: `npx prisma migrate resolve --applied 0_init`.
8. Commit the results.

ShouvikGhosh2048 commented on src/server/api/routers/user.ts:
```diff
@@ -203,6 +203,332 @@ export const userRouter = createTRPCRouter({
+                });
+            }
+
+            await ctx.prisma.publisher.updateMany({
```
Any suggestions on improving these 200 lines of foreign key changes? (Some of the inner objects are common so I guess I could create them once and then use them.)

otobot1 replied:
Yeah I added those common objects, added a check that the users in the claim actually exist, and wrapped it all in a transaction. The reason for the check that the users still exist, is so that if the db is ever in a slightly unexpected state, we don't make it worse by silently adding references to non-extant foreign keys. In most of the tRPC api (not sure if the oldest parts were done this way... i should check lol), we add foreign keys with the `{ connect: {...}}` syntax which automatically makes sure that the foreign keys are, in fact, valid. We can't use that here as `updateMany` doesn't support it, so we perform the check in our application instead and wrap it all in an [interactive transaction](https://www.prisma.io/docs/orm/prisma-client/queries/transactions#interactive-transactions).

ShouvikGhosh2048 commented on src/server/auth/prismaAdapter.ts:
```diff
@@ -11,7 +11,7 @@ type CreateUser = (user: PrismaUser) => Awaitable<TrimmedUser>;
 const getCreateUser = (prisma: PrismaClient): CreateUser => {
     return (user: PrismaUser): Awaitable<TrimmedUser> => {
-        const { id, ...rest } = user;   //remove the discord id from the user object - a replacement is auto-generated by Prisma
+        const { id, email, emailVerified, ...rest } = user;   //remove the discord id from the user object - a replacement is auto-generated by Prisma
```
Not type safe, but we need this to make auth work?

otobot1 replied:
Yeah it looks that way. It's probably fixable with a custom provider, but I don't think it's a very big deal.

otobot1 commented:
I should be able to take a look this weekend!

otobot1 committed:
tweak page URLs

combine user-claim related pages into single folder

otobot1 committed:
tweak wording on index page

make explicit that we don't want people duplicating their ratings

otobot1 committed:
Update schema.prisma

Tweak UserClaim column and relation names

otobot1 committed:
refactor checkIsPrivileged to support arrays of targetUserIds

otobot1 committed:
remove unused variable

otobot1 committed:
fix up tech and techVideo routers

otobot1 committed:
move userClaim pages into pages directory

required by next.js pagesRouter

otobot1 committed:
refactor userClaim api

backend only. frontend tweaks forthcoming.
-split userClaim endpoints into a subrouter
-place the user and userClaim router files into a new folder
-tweak userClaim api to make userClaims non-public, as they don't need to be public
-tweak userClaim router for formatting and readability, and make it match existing routers more closely
-wrap the userClaim verification db calls into a transaction

otobot1 committed:
Update userClaim.ts

-add check for claimedUser to verification endpoint

otobot1 committed:
create new baseline migration

otobot1 commented:
Hey @ShouvikGhosh2048 . I didn't quite manage to get through the entire thing this weekend, but I think it's in a state where you can keep working on it. I'm happy with the tRPC api now (although we still need server logging), but I didn't have a chance to really look at or adjust the frontend to match the backend changes, so I've left the frontend in a broken state (sorry!). I'll get to it when I can, and will mark this as a draft again when I do.

In the meantime, please feel free to work on the remaining pages or any other backend stuff. If you'd like to work on the pages you already created, please mark this PR as a draft and also make both pages work with the backend changes.

otobot1 committed:
show "claim user" message to all users

otobot1 committed:
Update /claim-user to work with API changes

Also make minor tweaks to wording

otobot1 committed:
increase query invalidation specificity

-update comment

otobot1 committed:
prefer import syntax: `type {...}` over `{type ...}`

-i read somewhere that it's better and found it convincing at the time, but i can't recall why. i'll add a comment to this PR (#861) if i ever remember

otobot1 committed:
refactoring, formatting, and referencing constants

otobot1 committed:
Update schema.prisma

change all `onUpdate: Restrict` to `onUpdate: Cascade`

otobot1 committed:
resolve webpack errors

otobot1 committed:
work on getting userClaim pages to load

no longer throwing React errors

otobot1 commented:
I made some progress on this tonight. I reviewed the new pages and I think I'm happy with how they work. I fixed all of the React errors and adjusted everything for the new API, but I'm still having permission errors that I haven't had a chance to figure out yet. I haven't been able to get either page to actually render on my end yet, so I may end up making some tweaks in the end. I also haven't had a chance to review the nextAuth changes, but I think any changes to that code will be minor.

@ShouvikGhosh2048 I'm done with my first-pass review, so I'm marking this as a draft again. To get this ready for merging, please:
1. Get both of the new pages working (sorry for breaking them lol).
2. Address the `TODO!!!` comments (only the ones with 3 exclamation marks).

I think that should be it, but I'm pretty tired at the moment. I may add another item or two in the next couple of days if I forgot something.

otobot1 committed:
get /claim-user working and tweak styling and wording

otobot1 committed:
get /claim-user/verify working

otobot1 committed:
rename /claim-user to /claim-users

otobot1 committed:
fix double call of layout component

otobot1 committed:
update comment

otobot1 committed:
update /claim-user pathname constants

otobot1 committed:
add link to verify page from claim page

only appears for admins

otobot1 committed:
address type error

otobot1 committed:
tidy up types and adjust comments

otobot1 committed:
Update schema.prisma

otobot1 committed:
resolve remaining TODOs

otobot1 committed:
remove remaining TODO!!! comments

otobot1 commented:
This PR is basically ready to go - there are just a couple of minor tweaks and a handful of follow-up issues to create. Whichever one of us has time first can work on the remaining changes - just communicate so that we aren't duplicating work.

Required Changes:
1. Move the user settings button.
  a. Remove the old button from the footer and move the remaining footer buttons so that the footer is visually balanced.
  b. Create a new user settings button next to the `Login` button.
    - The new button should still link to `/coming-soon`.
    - The button should be a gear icon.
2. Style the `CLAIM YOUR OLD USER` link on the home page so that it's clear that it is clickable.
  a. Make the cursor change on hover to indicate that the link is clickable.
3. Add server logging to all mutations in the UserClaim router.

Follow-Up Issues:
1. Add server-side logic to the sign-in flow to check the user's `accountStatus` and return an error for non-active users.
  a. This will be required for v0.2.0.
2. Remove unused procedures from the User router.
3. Refactor the `CLAIM YOUR OLD USER` link on the home page that logged-out users see, and the `Login` button, to show a destination URL on hover like a normal link.
  a. This is low priority.
  b. If we need to fake this somehow, that's fine.
4. Use the new `satisfies` keyword to explicitly link the `selectObject` constants and associated types in each tRPC router.
  a. See the `RelatedUser` type and `relatedUserSelectObject` constant in the UserClaim router.
5. Determine if NextAuth is requesting the users' emails from Discord for user creation and/or sign-in and then discarding the info, or if we don't request it at all.
  a. High priority.
  b. We will need to include this info in our privacy policy.

otobot1 committed:
remove old settings button and rearrange remaining footer links

otobot1 committed:
cleanup header JSX

otobot1 committed:
add `as const`

otobot1 committed:
finish cleaning up header JSX and fixing CSS

otobot1 committed:
add more whitespace

otobot1 committed:
add new settings button

otobot1 commented:
I'm working on this now. I should be able to finish up this PR and get it merged soon.

otobot1 committed:
Update schema.prisma

make user [discordUsername, discordDiscriminator] index non-unique to avoid errors when creating new accounts for legacy users

otobot1 commented:
> This PR is basically ready to go - there are just a couple of minor tweaks and a handful of follow-up issues to create. Whichever one of us has time first can work on the remaining changes - just communicate so that we aren't duplicating work.
> 
> Required Changes:
> 
> 1. Move the user settings button.
>    a. Remove the old button from the footer and move the remaining footer buttons so that the footer is visually balanced.
>    b. Create a new user settings button next to the `Login` button.
>    
>    * The new button should still link to `/coming-soon`.
>    * The button should be a gear icon.
> 2. Style the `CLAIM YOUR OLD USER` link on the home page so that it's clear that it is clickable.
>    a. Make the cursor change on hover to indicate that the link is clickable.
> 3. Add server logging to all mutations in the UserClaim router.
> 
> Follow-Up Issues:
> 
> 1. Add server-side logic to the sign-in flow to check the user's `accountStatus` and return an error for non-active users.
>    a. This will be required for v0.2.0.
> 2. Remove unused procedures from the User router.
> 3. Refactor the `CLAIM YOUR OLD USER` link on the home page that logged-out users see, and the `Login` button, to show a destination URL on hover like a normal link.
>    a. This is low priority.
>    b. If we need to fake this somehow, that's fine.
> 4. Use the new `satisfies` keyword to explicitly link the `selectObject` constants and associated types in each tRPC router.
>    a. See the `RelatedUser` type and `relatedUserSelectObject` constant in the UserClaim router.
> 5. Determine if NextAuth is requesting the users' emails from Discord for user creation and/or sign-in and then discarding the info, or if we don't request it at all.
>    a. High priority.
>    b. We will need to include this info in our privacy policy.

Required Item 1 is done. The verify users transaction broke at some point though. I'll have to fix that.

otobot1 committed:
revert dad1ec0d9a56c5ba57528811a8c09fe66cec7358

otobot1 committed:
start troubleshooting verify user transaction
